snow/engine/snowman/engine_decorator.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package snowman
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+)
+
+type EngineStragglerDetector struct {
+	Listener func(duration time.Duration)
+	Time     func() time.Time
+}
+
+func (ed *EngineStragglerDetector) AttachToEngine(e *Engine) common.Engine {
+	minConfRatio := float64(e.Params.AlphaConfidence) / float64(e.Params.K)
+	sd := newStragglerDetector(ed.Time, e.Config.Ctx.Log, minConfRatio, e.Consensus.LastAccepted,
+		e.Config.ConnectedValidators.ConnectedValidators, e.Config.ConnectedValidators.ConnectedPercent,
+		e.Consensus.Processing, e.acceptedFrontiers.LastAccepted)
+	de := &DecoratedEngine{Engine: e}
+	de.decorate("Chits", func(e *Engine) {
+		behindDuration := sd.CheckIfWeAreStragglingBehind()
+		if behindDuration > 0 {
+			e.Config.Ctx.Log.Info("We are behind the rest of the network", zap.Float64("seconds", behindDuration.Seconds()))
+		}
+		e.metrics.stragglingDuration.Set(float64(behindDuration))
```

Comment:
kind of subjective but it seems nicer if these were handled in the Listener (doesn't seem the listener is used?)

Reply:
The listener is not used yet, it is preparation for a future PR (which I don't have yet). 
However, after your recommendation to unify the decorator and the detector, there is no listener anymore.